File: inject/java/cat/narezany/margyt/Accent.java
package cat.narezany.margyt;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.drawable.GradientDrawable;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bytedance.tux.icon.TuxIconView;

/**
 * TikTok's accent colour, made changeable.
 *
 * The app is built around one pink, `#FE2C55`: the like, the follow button, the
 * tab underline, the badges. Most of the places it is drawn hold it as a plain
 * constant in the bytecode -- `const v1, -0x1d3ab` before a `Paint.setColor` --
 * so the build rewrites each of those into a call here, and the colour becomes
 * whatever this returns.
 *
 * This sits in the drawing path of half the app, so it answers from a cached
 * int and never throws: the worst it can do when something is wrong is hand
 * back the pink the app came with.
 */
public final class Accent {

    private Accent() {}

    /**
     * TikTok's own pink, and always the colour a swap is measured from.
     *
     * It stays the reference even when the build baked something else into the
     * resources. Making the baked colour the reference instead is a mistake
     * that was made once and is worth writing down: the app then stops
     * recognising its own pink wherever the build did not reach -- which is
     * most of the bytecode -- and starts recognising whatever sits near the
     * baked colour by hue, which for a mint is the green of somebody being
     * online. Shades this build wrote are recognised by the list in Baked
     * instead, exactly, and sent wherever their original would go.
     */
    public static final int TIKTOK = 0xFFFE2C55;

    /** What this apk was built with: the colour before anyone chooses another. */
    public static final int BUILT_WITH = Baked.ACCENT;

    public static final String KEY = "accent";

    /** iso-style names are not needed here; the label is the colour itself. */
    public static final int[] PALETTE = {
            BUILT_WITH,
            TIKTOK,      // TikTok's own pink, in case the build baked another
            0xFF8DD1B0,  // Margy mint
            0xFF25F4EE,  // TikTok's own cyan
            0xFF4C8DFF,
            0xFF9B6BFF,
            0xFFFF8A3D,
            0xFF35C759,
            0xFFFFD23F,
            0xFFFF4D6D,
            0xFFE8E8E8,
    };

    /** The palette without repeats: the built-in colour may be in it twice. */
    public static int[] palette() {
        int[] out = new int[PALETTE.length];
        int count = 0;
        for (int colour : PALETTE) {
            boolean seen = false;
            for (int i = 0; i < count; i++) seen |= out[i] == colour;
            if (!seen) out[count++] = colour;
        }
        int[] trimmed = new int[count];
        System.arraycopy(out, 0, trimmed, 0, count);
        return trimmed;
    }

    private static volatile int cached;

    public static int colour() {
        int known = cached;
        if (known != 0) return known;
        SharedPreferences prefs = prefs();
        if (prefs == null) return BUILT_WITH;  // too early to know; do not cache it
        int chosen = BUILT_WITH;
        try {
            chosen = prefs.getInt(KEY, BUILT_WITH);
        } catch (Throwable ignored) {
        }
        if (chosen == 0) chosen = BUILT_WITH;
        cached = chosen;
        return chosen;
    }

    /**
     * Where a rewritten constant lands.
     *
     * `const v1, -0x1d3ab` in TikTok's bytecode becomes a call to this, so it
     * is the accent as chosen -- and then whatever the plugins make of it.
     * `colour()` itself stays plain: it is what the mod paints its own screen
     * with, and a plugin recolouring the settings it is being configured from
     * would be a poor joke.
     */
    public static int accent() {
        return Plugins.colour(colour());
    }

    public static void set(int colour) {
        cached = colour == 0 ? BUILT_WITH : colour;
        SharedPreferences prefs = prefs();
        if (prefs != null) prefs.edit().putInt(KEY, cached).apply();
    }

    public static boolean isDefault() {
        return colour() == BUILT_WITH;
    }

    private static SharedPreferences prefs() {
        try {
            Context context = Margy.context();
            if (context == null) return null;
            return context.getSharedPreferences(Margy.PREFS, Context.MODE_PRIVATE);
        } catch (Throwable ignored) {
            return null;
        }
    }

    // ------------------------------------------------- where colours arrive

    /**
     * Any colour of TikTok's red family comes back on the chosen accent.
     *
     * Not the one value it used to be: the app draws its pink at a dozen
     * opacities and next to a family of neighbours, and swapping only the exact
     * brand colour left nine tenths of the red on screen. Palette says what
     * belongs to the family and where it moves to.
     */
    public static int swap(int colour) {
        // Paint.setColor is called on every frame of everything, so the answer
        // to the question just asked is kept: one long holds both halves, so a
        // reader either sees a whole pair or none of it, with no lock either way
        long known = memo;
        if ((int) (known >>> 32) == colour) return (int) known;

        int out = Plugins.colour(translate(colour));
        memo = ((long) colour << 32) | (out & 0xFFFFFFFFL);
        return out;
    }

    /**
     * Two ways in, and the second only when the first says nothing.
     *
     * TikTok's own family is recognised by hue, which reaches every shade of
     * it including the ones no build ever saw. What the build baked is not a
     * family at all -- it is a list -- and each entry knows the shade it was
     * made from, so it is sent wherever that shade would go now.
     */
    private static int translate(int colour) {
        int chosen = colour();
        int moved = Palette.map(colour, TIKTOK, chosen);
        if (moved != colour) return moved;

        int origin = originOf(colour);
        if (origin != 0) return Palette.map(origin, TIKTOK, chosen);

        // last, and only for the colours TikTok repaints itself when its own
        // theme changes: the accent has had its say and did not want this one
        return Themes.recolour(colour);
    }

    /** The shade a baked colour was made from, or zero. Binary search. */
    private static int originOf(int colour) {
        int[] baked = Baked.BAKED;
        int low = 0, high = baked.length - 1;
        while (low <= high) {
            int middle = (low + high) >>> 1;
            int here = baked[middle];
            if (here == colour) return Baked.FROM[middle];
            if (here < colour) low = middle + 1; else high = middle - 1;
        }
        return 0;
    }

    private static volatile long memo;

    /** Forget the one remembered answer: something it depended on has changed. */
    static void forget() {
        memo = 0;
    }

    // ---------------------------------------------- where a colour is used

    /**
     * The other half of the accent.
     *
     * Reading a colour is not the only way to have one: it can be computed,
     * blended, or carried in from somewhere the mod never sees. But it has to
     * be applied to something before it is drawn, and there are only so many
     * ways to apply one. Every call site of these in TikTok's bytecode is
     * rewritten to come through here, which is how the accent reaches what the
     * resource table and the constants never could.
     */
    public static void setColor(Paint paint, int colour) {
        paint.setColor(swap(colour));
    }

    public static void setColor(GradientDrawable shape, int colour) {
        shape.setColor(swap(colour));
    }

    public static void setColor(TuxIconView icon, int colour) {
        icon.setColor(swap(colour));
    }

    public static void setColorFilter(ImageView view, int colour) {
        view.setColorFilter(swap(colour));
    }

    public static void setColorFilter(ImageView view, int colour, PorterDuff.Mode mode) {
        view.setColorFilter(swap(colour), mode);
    }

    public static void setTextColor(TextView view, int colour) {
        view.setTextColor(swap(colour));
    }

    public static void setBackgroundColor(View view, int colour) {
        view.setBackgroundColor(swap(colour));
    }

    public static ColorStateList valueOf(int colour) {
        return ColorStateList.valueOf(swap(colour));
    }

    public static int getColor(Context context, int id) {
        return swap(context.getColor(id));
    }

    public static int getColor(Resources resources, int id) {
        return swap(resources.getColor(id));
    }

    public static int getColor(Resources resources, int id, Resources.Theme theme) {
        return swap(resources.getColor(id, theme));
    }

    public static int getColor(TypedArray array, int index, int fallback) {
        return swap(array.getColor(index, fallback));
    }
}

File: tiktokmi/accent.py
"""The accent colour, baked into the places code cannot reach.

Most of TikTok's pink is reachable at runtime: a constant in the bytecode, or a
colour resource read through `Resources.getColor`, both of which the dex patch
sends through the mod. What is not reachable is the pink written into compiled
XML -- a vector's `fillColor`, a selector's item, a shape's solid -- because
the framework parses those itself, inside its own code, where nothing of ours
runs.

Those are settled here instead, while the apk is being built, and both edits
are the size-preserving kind: a colour is eight bytes of Res_value wherever it
appears, so the new one goes exactly where the old one was.
"""

from __future__ import annotations

from typing import Dict, List, Tuple

from .apkzip import Apk
from .arsc import Arsc
from .axml import Axml
from .palette import captures, map_colour

# the colour types a Res_value can be: argb8, rgb8, argb4, rgb4
COLOUR_TYPES = (0x1C, 0x1D, 0x1E, 0x1F)


def bake(apk: Apk, arsc: Arsc, old: int, new: int, pairs: Dict[int, int] = None) -> List[str]:
    """Move TikTok's red family onto the accent, in the table and in compiled XML.

    Not one value: the family, at every opacity it is used at, mapped by the
    step that takes the brand pink to the chosen colour. tiktokmi/palette.py says
    what is taken over and what is left alone.

    Asked for the colour the apk already has, this counts instead of writing --
    and says what it counted. Those places are the ones the bytecode patch can
    never reach, because the framework parses them inside its own code, so a
    build that leaves them alone should say so rather than report nothing to do.
    """
    writing = old != new
    report = []
    if pairs is None:
        pairs = {}

    entries, shades = _walk_table(arsc, old, new, writing, pairs)

    files = 0
    changes = 0
    for name in apk.names():
        if not (name.startswith("res/") and name.endswith(".xml")):
            continue
        try:
            axml = Axml.parse(apk.read(name))
        except Exception:
            continue
        here = 0
        for node in axml.nodes:
            for attribute in node.attributes:
                if attribute.kind in COLOUR_TYPES and captures(attribute.data, old):
                    shades.add(attribute.data)
                    if writing:
                        moved = map_colour(attribute.data, old, new)
                        # what the app will now hold, and what it used to be:
                        # the runtime needs the original to know where to send it
                        pairs[moved] = attribute.data
                        attribute.data = moved
                    here += 1
        if here:
            if writing:
                apk.replace(name, axml.build())
            files += 1
            changes += here

    if writing:
        report.append("resource entries: %d" % entries)
        report.append("compiled xml: %d colours in %d files" % (changes, files))
        report.append("%d shades of the family in all, alpha kept, moved together"
                      % len(shades))
    else:
        report.append("the accent is TikTok's own, nothing to bake")
        report.append("%d in the table and %d in %d compiled xml files stay as they are"
                      % (entries, changes, files))
        report.append("those are pictures rather than code -- the palette in the "
                      "settings cannot reach them; --accent is what moves them")
    return report


def _walk_table(arsc: Arsc, old: int, new: int, writing: bool, pairs: Dict[int, int]):
    """Every typed value in the table that belongs to the family."""
    found = 0
    shades = set()
    for package in arsc.packages:
        for type_id in list(package.types):
            for entry in _entries(arsc, package, type_id):
                if entry.kind in COLOUR_TYPES and captures(entry.data, old):
                    shades.add(entry.data)
                    if writing:
                        moved = map_colour(entry.data, old, new)
                        pairs[moved] = entry.data
                        arsc.set_value(entry, entry.kind, moved)
                    found += 1
    return found, shades


def _entries(arsc: Arsc, package, type_id: int):
    """Every value of every entry of one type, across its configurations."""
    for entry_id in range(_entry_count(arsc, package, type_id)):
        res_id = (package.id << 24) | (type_id << 16) | entry_id
        for value in arsc.values(res_id):
            yield value


def _entry_count(arsc: Arsc, package, type_id: int) -> int:
    import struct

    most = 0
    for chunk in package.types.get(type_id, []):
        count = struct.unpack_from("<I", arsc.data, chunk + 12)[0]
        most = max(most, count)
    return most
